lib/xdparser2.ts

```diff
@@ -674,94 +674,130 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
   return splits
 }
 
-export function inlineMarkdownParser(str: string): MDClueComponent[] {
-  const components: MDClueComponent[] = []
-  let token = ""
-  let mode: MDClueComponent[0] = "text"
-  let linkText = ""
+function isAlphabetical(str: string): boolean {
+  return str.charCodeAt(0) >= 97 && str.charCodeAt(0) <= 122
+}
 
-  const pushText = () => {
-    if (token.length > 0) {
-      components.push(["text", token])
-      token = ""
-    }
+function parseBBTagType(str: string): string {
+  let index = 0
+  let tagType = ""
+  while (index < str.length && isAlphabetical(str[index])) {
+    tagType += str[index]
+    ++index
   }
+  return tagType
+}
 
-  for (let index = 0; index < str.length; index++) {
-    const prevLetter = index > 0 ? str.slice(index - 1, index) : ""
-    const letter = str.slice(index, index + 1)
-    const nextLetter = str.slice(index + 1, index + 2)
+// parses the inner url starting tags
+function parseURL(str: string): string {
+  let index = 0
+  // whitespace is allowed between [url and =
+  while (index < str.length && str[index] === " ") {
+    ++index
+  }
+  // checks to see if the token is '='
+  if (str[index] !== "=") {
+    return ""
+  }
+  ++index
 
-    const otherMDKeys = ["*", "/", "~", "["]
-    if (letter === "\\" && otherMDKeys.includes(nextLetter)) {
-      index++
-      token += nextLetter
-      continue
-    } else if (mode === "text" && prevLetter !== "\\") {
-      if (letter === "[") {
-        mode = "link"
-        pushText()
-        continue
-      }
+  // whitespace is allowed between '=' and the actual url
+  while (index < str.length && str[index] === " ") {
+    ++index
+  }
 
-      // Bold is started by two asterisks
-      if (letter === "*" && nextLetter === "*") {
-        mode = "bold"
-        // Jump a letter
-        index++
-        pushText()
-        continue
-      }
+  let url = ""
+  while (index < str.length && str[index] !== "]") {
+    url += str[index]
+    ++index
+  }
+  return url
+}
 
-      if (letter === "/" && prevLetter === " ") {
-        mode = "italics"
-        pushText()
-        continue
-      }
+function parseBBInnerContents(str: string) {
+  let index = 0;
+  let start = index;
+  while (index < str.length && str[index] !== "[") {
+    ++index
+  }
+  return str.slice(start, index)
+}
 
-      if (letter === "~") {
-        mode = "strike"
-        pushText()
-        continue
+function parseBB(str: string): { mdComponent: MDClueComponent; str: string } | undefined {
+  let index = 1 // starting at one because [ is the start
+  let starting = 0 // used for keeping the start of the entire bb tag/group
+  let tagType = parseBBTagType(str.slice(index))
+  if (tagType === "b" || tagType === "url" || tagType === "i" || tagType === "s") {
+    index += tagType.length
+    let url;
+    if (tagType === "url") {
+      url = parseURL(str.slice(index))
+      if (url.length === 0) {
+        // no url so invalid tag
+        return
```

Comment:
What happens if parseBB returns undefined exactly? Could be more explicit to throw an error and handle the error in the caller.

Reply:
we could throw but something failing to parse as BBCode would be just added as ur typical text so 🤷